File: src/main/java/no/capra/AppMain.java
package no.capra;

import no.capra.config.BasicAuthSecurityHandler;
import no.capra.config.JerseyConfig;
import no.capra.config.JsonJettyErrorHandler;
import no.capra.config.PropertiesHelper;
import no.capra.health.HealthEndpoint;
import org.eclipse.jetty.security.SecurityHandler;
import org.eclipse.jetty.server.Server;
import org.eclipse.jetty.servlet.ServletContextHandler;
import org.eclipse.jetty.servlet.ServletHolder;
import org.glassfish.jersey.servlet.ServletContainer;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.Properties;

public class AppMain {
    private static final Logger log = LoggerFactory.getLogger(AppMain.class);

    static final String CONTEXT_PATH = System.getProperty("server.context.path", "/");
    private final Integer port;
    private final Properties properties;
    private Server server;

    AppMain(Integer port, Properties properties) {
        this.port = port;
        this.properties = properties;
    }

    public static void main(String[] args) throws Exception {
        Integer port = Integer.parseInt(System.getProperty("server.port", "8080"));
        Properties properties = PropertiesHelper.getProperties();
        new AppMain(port, properties).start();
        log.info("Server stopped");
    }

    void start() throws InterruptedException {
        log.debug("Starting server at port {}", port);
        server = new Server(port);
        server.setHandler(getServletContextHandler());
        server.setStopAtShutdown(true);

        try {
            server.start();
        } catch (Exception e) {
            log.error("Error during Jetty startup. Exiting", e);
        }
        String healthEndpoint = "http://localhost:" + getPort() + HealthEndpoint.HEALTH_PATH;
        log.info("Server started at port {}. Health endpoint at: {}", port, healthEndpoint);
        server.join();
    }

    private ServletContextHandler getServletContextHandler() {
        ServletContextHandler contextHandler = new ServletContextHandler();
        contextHandler.setContextPath(CONTEXT_PATH);

        String username = PropertiesHelper.getStringProperty(properties, PropertiesHelper.BASIC_AUTH_USERNAME, null);
        String password = PropertiesHelper.getStringProperty(properties, PropertiesHelper.BASIC_AUTH_PASSWORD, null);

        // Basic Authentication
        SecurityHandler basicAuthSecurityHandler = BasicAuthSecurityHandler
                .getBasicAuthSecurityHandler(username, password, "realm");
        contextHandler.setSecurityHandler(basicAuthSecurityHandler);

        // Add Jersey servlet to the Jetty context
        ServletHolder jerseyServlet = new ServletHolder(new ServletContainer(new JerseyConfig(properties)));
        contextHandler.addServlet(jerseyServlet, "/*");

        // Error responses as application/json with proper charset
        contextHandler.setErrorHandler(new JsonJettyErrorHandler());

        return contextHandler;
    }

    //used by TestServer
    void stop() {
        try {
            server.stop();
        } catch (Exception e) {
            log.warn("Error while stopping Jetty server", e);
        }
    }

    Integer getPort() {
        return port;
    }

    boolean isStarted() {
        return server != null && server.isStarted();
    }


}

File: src/main/kotlin/no/capraconsulting/AppMain.kt
package no.capraconsulting

import no.capraconsulting.config.BasicAuthSecurityHandler
import no.capraconsulting.config.JerseyConfig
import no.capraconsulting.config.JsonJettyErrorHandler
import no.capraconsulting.config.PropertiesHelper
import no.capraconsulting.config.PropertiesHelper.requireString
import no.capraconsulting.health.HealthEndpoint
import org.eclipse.jetty.server.Server
import org.eclipse.jetty.servlet.ServletContextHandler
import org.eclipse.jetty.servlet.ServletHolder
import org.glassfish.jersey.servlet.ServletContainer
import org.slf4j.LoggerFactory
import java.util.Properties
import kotlin.system.exitProcess

private val log = LoggerFactory.getLogger(AppMain::class.java)

fun main() {
    val port = Integer.parseInt(System.getProperty("server.port", "8080"))
    AppMain(port, PropertiesHelper.getProperties()).start()
    log.info("Server stopped")
}

class AppMain(val port: Int, private val properties: Properties) {
    private var server: Server? = null

    @Throws(InterruptedException::class)
    fun start() {
        log.debug("Starting server at port {}", port)

        val server = Server(port)
        this.server = server

        server.handler = getServletContextHandler()
        server.stopAtShutdown = true

        try {
            server.start()
        } catch (e: Exception) {
            log.error("Error during Jetty startup. Exiting", e)
            exitProcess(1)
        }

        val healthEndpoint = "http://localhost:" + port + HealthEndpoint.HEALTH_PATH
        log.info("Server started at port {}. Health endpoint at: {}", port, healthEndpoint)
        server.join()
    }

    private fun getServletContextHandler(): ServletContextHandler {
        val contextHandler = ServletContextHandler()
        contextHandler.contextPath = CONTEXT_PATH

        val username = properties.requireString(PropertiesHelper.BASIC_AUTH_USERNAME)
        val password = properties.requireString(PropertiesHelper.BASIC_AUTH_PASSWORD)

        // Basic Authentication
        contextHandler.securityHandler = BasicAuthSecurityHandler
            .getBasicAuthSecurityHandler(username, password, "realm")

        // Add Jersey servlet to the Jetty context
        val jerseyServlet = ServletHolder(ServletContainer(JerseyConfig(properties)))
        contextHandler.addServlet(jerseyServlet, "/*")

        // Error responses as application/json with proper charset
        contextHandler.errorHandler = JsonJettyErrorHandler()

        return contextHandler
    }

    // used by TestServer
    fun stop() {
        try {
            server?.stop()
        } catch (e: Exception) {
            log.warn("Error while stopping Jetty server", e)
        }
    }

    fun isStarted(): Boolean {
        return server?.isStarted == true
    }

    companion object {
        @JvmField
        val CONTEXT_PATH: String = System.getProperty("server.context.path", "/")
    }
}
